pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -0,0 +1,160 @@
+    from mcp.types import CallToolResult
+except ImportError as _import_error:
+    raise ImportError(
+        'Please install the `mcp` package to use the MCP server, '
+        "you can use the `mcp` optional group — `pip install 'pydantic-ai-slim[mcp]'`"
+    ) from _import_error
+
+__all__ = ('MCPServer', 'MCPSubprocessServer', 'MCPRemoteServer')
+
+
+class MCPServer(ABC):
+    """Base class for MCP servers that can be used to run a command or connect to an SSE server.
+
+    See <https://modelcontextprotocol.io/introduction> for more information.
+    """
+
+    is_running: bool = False
+
+    _client: ClientSession
+    _read_stream: MemoryObjectReceiveStream[JSONRPCMessage | Exception]
+    _write_stream: MemoryObjectSendStream[JSONRPCMessage]
+    _exit_stack: AsyncExitStack
+
+    @abstractmethod
+    @asynccontextmanager
+    async def client_streams(
+        self,
+    ) -> AsyncIterator[
+        tuple[MemoryObjectReceiveStream[JSONRPCMessage | Exception], MemoryObjectSendStream[JSONRPCMessage]]
+    ]:
+        """Create the streams for the MCP server."""
+        raise NotImplementedError('MCP Server subclasses must implement this method.')
+        yield
+
+    async def list_tools(self) -> list[ToolDefinition]:
+        """Retrieve tools that are currently active on the server.
+
+        Note:
+        - We don't cache tools as they might change.
+        - We also don't subscribe to the server to avoid complexity.
+        """
+        tools = await self._client.list_tools()
+        return [
+            ToolDefinition(
+                name=tool.name,
+                description=tool.description or '',
+                parameters_json_schema=tool.inputSchema,
+            )
+            for tool in tools.tools
+        ]
+
+    async def call_tool(self, tool_name: str, arguments: dict[str, Any]) -> CallToolResult:
+        """Call a tool on the server.
+
+        Args:
+            tool_name: The name of the tool to call.
+            arguments: The arguments to pass to the tool.
+
+        Returns:
+            The result of the tool call.
+        """
+        return await self._client.call_tool(tool_name, arguments)
+
+    async def __aenter__(self) -> MCPServer:
+        self._exit_stack = AsyncExitStack()
+
+        self._read_stream, self._write_stream = await self._exit_stack.enter_async_context(self.client_streams())
+        client = ClientSession(read_stream=self._read_stream, write_stream=self._write_stream)
+        self._client = await self._exit_stack.enter_async_context(client)
+
+        await self._client.initialize()
+        self.is_running = True
+        return self
+
+    async def __aexit__(
+        self, exc_type: type[BaseException] | None, exc_value: BaseException | None, traceback: TracebackType | None
+    ) -> bool | None:
+        await self._exit_stack.aclose()
+        self.is_running = False
+
+
+@dataclass
+class MCPSubprocessServer(MCPServer):
```

Comment:
I think I was wrong about this name, we should use the names MCP users which are clearly "stdio" and "sse", my mistake ✋ .

I guess we should call this `MCPServerStdio`, and the other one `MCPServerSSE`.

Reply:
> I think I was wrong about this name, we should use the names MCP users which are clearly "stdio" and "sse", my mistake ✋ .
> 
> I guess we should call this `MCPServerStdio`, and the other one `MCPServerSSE`.

...